### PR TITLE
Update Transaction History

### PR DESCRIPTION
**Pull request by rickycodes**
re: https://github.com/MetaMask/metamask-extension/issues/7317
re: https://github.com/MetaMask/Design/issues/139

I've set up storybook here: https://rickycodes.github.io/metamask-extension/

The new bits are the `Item`'s and `PopOver`'s

there's still a few things remaining, which I will list out here, but hopefully we can get some :eyes: on this and people can see progress or chime in on anything that looks _way_ off

### Items remaining:
- [x] <strike>clean up `JSX` structure and continue breaking things out into smaller pure components</strike>
- [x] <strike>clean up scss `classNames` and structure (use existing components as reference)</strike>
- [x] list styles for `popover-approve` need some love (in particular the transaction history list-items)

### CODE REVIEW

File: ui/app/components/ui/icon/icon.component.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+const Icon = ({ className, width, height, color, type }) => {
+  const svgs = {
+    send: (
+      <svg className={className} width={width} height={height} viewBox="0 0 30 30" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (danjm) [inline ui/app/components/ui/icon/icon.component.js]: Why did you put the svgs here instead of in `.svg` files in the `images/` directory?

File: ui/app/components/ui/item/item.component.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  cash,
+}) => {
+  const isApproved = status === 'approved'
+  const isPending = status === 'pending'
+  const isFailed = status === 'failed'
+
+  if (status === 'unapproved') {
+    subtitle = (
+      <h3>
+        <span className="unapproved">Unapproved</span> · {subtitle}
```
Comment (danjm) [inline ui/app/components/ui/item/item.component.js]: translate strings in this file

File: ui/app/components/ui/item/item.component.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+
```
Comment (danjm) [inline ui/app/components/ui/item/item.component.js]: will these items always be items in a list? if so, the code might be a little more self-documenting if we named this "list-item"

File: ui/app/components/ui/item/item.component.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  title: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.node,
+  ]).isRequired,
+  subtitle: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.node,
+  ]).isRequired,
+  more: PropTypes.node,
+  crypto: PropTypes.string,
```
Comment (danjm) [inline ui/app/components/ui/item/item.component.js]: The prop names "crypto" and "cash" are not quite reflective of what is possible in the extension. Users can set their desired currency conversions on the settings screen, and then select whether they want eth or the selected one to be primary. This makes it possible for both currencies to be cryptos.

See screenshots

Screenshot from 2020-01-15 11-51-58
Screenshot from 2020-01-15 11-50-09

In other places in we've used "nativeCurrency" for ETH and "currentCurrency" for whatever the user has selected.

File: ui/app/css/itcss/components/item.scss [hunk 5]
```diff
@@ -0,0 +1,116 @@
+.Item__flex-grid {
```
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: Nit: most class names in the code-base are not capitalized. Lower case `I` for consistency?

File: ui/app/css/itcss/components/item.scss [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  width: 100%;
+  min-width: 374px;
```
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: Will need some additional css to accomodate the popup view, which is 357px wide. Perhaps a media query.

File: ui/app/css/itcss/components/item.scss [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  min-height: 86px;
+  margin: 0 20px;
+
+  left: 0px;
+  top: 497px;
```
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: Not sure that these `left` and `top` rules are needed

File: ui/app/css/itcss/components/item.scss [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  background: #FFFFFF;
+  padding: 24px 16px;
+  font-family: 'Roboto', sans-serif;
+  font-style: normal;
+  font-weight: normal;
+  border-top: 1px solid #E5E5E5;
```
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: In the interest of stepping towards a design system, can we add a variable for `#E5E5E5` to `variables.scss` and us the variable here instead? Our names for these variable has been arbitrary in the past (using http://chir.ag/projects/name-that-color/). We could now start using some sort of functional names.

File: ui/app/css/itcss/components/item.scss [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    // these will need to be broken out into other components
+    .more {
+      button {
+        border: 1px solid #6A737D;
+        color: #6A737D;
+        box-sizing: border-box;
+        border-radius: 4px;
+        background: transparent;
+        font-family: 'Roboto', sans-serif;
```
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: Nit: we can use the `%font` mixin here instead of font-family, style and weight

File: ui/app/css/itcss/components/item.scss (hunk 5 above)
Comment (danjm) [inline ui/app/css/itcss/components/item.scss]: We usually place scss files meant for usage with one component in the same directory as the components js files. I'd recommend that here, unless you had a reason not to.

File: ui/app/css/itcss/components/popover.scss [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+  box-shadow: 0px 4px 30px rgba(0, 0, 0, 0.25);
+  border-radius: 10px;
+  background: white;
+  z-index: 2;
+  left: calc(50% - 328px/2);
+  top: 10px;
+
+  .header {
+    h2 {
+      font-family: Roboto;
```
Comment (danjm) [inline ui/app/css/itcss/components/popover.scss]: Nite: See above comment on font mixins. The font mixins are added in `variables.scss`. An example usage is https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/components/app/permission-page-container/index.scss (`@extend...`)

File: ui/app/css/itcss/components/popover.scss [hunk 11]
```diff
@@ -0,0 +1,328 @@
+
```
Comment (danjm) [inline ui/app/css/itcss/components/popover.scss]: Are you familiar with BEM css style? We try to mostly follow it throughout the codebase, and something like it will be important as we build out a design system. Intro and further links are here: http://getbem.com/introduction/

I think it is best to change the naming here to reflect that. (Sorry, I know this is an annoying request)

File: ui/app/components/ui/transaction-details/transaction-details.component.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+const TransactionDetails = ({
+  amount,
+  fee,
+  ethCost,
+}) => (
+  <div className="details">
+    <div className="details-amounts">
+      <div className="col">
+        Amount<br />
+        Transaction Fee
```
Comment (danjm) [inline ui/app/components/ui/transaction-details/transaction-details.component.js]: translate the strings in this file

**Review by danjm — CHANGES_REQUESTED** [review]
Completed an initial code review. Left some comments. Happy to discuss any of them further.

File: ui/app/components/ui/icon/icon.component.js (hunk 1 above)
Comment (rickycodes) [reply ui/app/components/ui/icon/icon.component.js]: my thinking is that it's faster to render the svgs directly inline (it's basically just template data anyway). this means that there's no http request for svgs as image assets (faster renders since we're removing that overhead)

however, I realise the downside of this is that updating these assets becomes a bit more complicated (though, the "copy as svg text" functionality in figma makes it only slightly more so?)

Comment (rickycodes) [reply ui/app/components/ui/icon/icon.component.js]: additionally this lets us modify the svg's fill prop easily to change an SVG colour... otherwise you'd need multiple assets for each colour treatment

File: ui/app/components/ui/popover/index.scss [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+  &-container {
+    padding: 0 5px 0 16px;
+    margin-top: 5px;
+    margin-bottom: -10px;
+    margin-right: 5px;
+    height: calc(100% - 66px - 10px);
+    overflow-y: scroll;
+    position: relative;
+    display: flex;
+    // align-items: stretch;
```
Comment (danjm) [inline ui/app/components/ui/popover/index.scss]: nit: delete commented out code

File: ui/app/components/ui/popover/popover.stories.js [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+const preventDefault = (event) => {
+  event.preventDefault()
+}
+
+const Chart = () => (
+  <div className="popover__chart"></div>
+)
+
+storiesOf('PopOver', module)
+  .add('PopOver - Approve', () => (
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/popover.stories.js]: These stories seem to depict the transactions details popover specifically, rather than the base popover component. I would expect these stories to be simpler - focused on demonstrating what the popover can do, how it scales to different sizes, etc.

Maybe these could stay here as complex examples, but they seem more appropriate to a `TransactionDetailsPopover` component perhaps?

File: ui/app/components/ui/popover/popover.component.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import PopOverheader from './popover.header.component'
+
+const PopOver = ({ title, children, onClose }) => (
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/popover.component.js]: Nit: Why `PopOver` instead of `Popover`? I thought the word was "popover".

File: ui/app/components/ui/popover/popover.header.component.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+const PopOverheader = ({ title, onClose }) => (
+  <div className="popover-header">
+    <h2 className="popover-header__heading">{title}</h2>
+    <a href="#" className="popover-header__close" onClick={onClose}>
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/popover.header.component.js]: Could this be made a button instead of an anchor? It seems like more of a button than an anchor to me.

File: ui/app/components/ui/popover/popover.header.component.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+      <div className="popover-header__close--wrap">
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/popover.header.component.js]: Nit: `wrap`, `line-a`, and `line-b` are not modifiers - they're elements. So the proper BEM classname would be `popover-header__close-wrap` here, for example

File: ui/app/components/ui/popover/popover.component.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+      {children}
+    </div>
+    <a href="#" className="popover-bg" onClick={onClose} />
+  </div>
+)
+
+PopOver.propTypes = {
+  title: PropTypes.string.isRequired,
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/popover.component.js]: This PropType is redundant because an array of nodes is also a node.

I noticed this pattern already existed in the codebase, so I've removed it in #8126

File: ui/app/components/ui/popover/index.scss [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    top: 0;
+    left: 0;
+    width: 100%;
+    height: 100%;
+    background: black;
+    opacity: 20%;
+    z-index: 1;
+  }
+
+  &__advanced-input-controls {
```
Comment (Gudahtt) [inline ui/app/components/ui/popover/index.scss]: A lot of these styles appear unrelated to the popover component

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: ui/app/components/ui/icon/icon.component.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        <path fillRule="evenodd" clipRule="evenodd" d="M15 29C22.732 29 29 22.732 29 15C29 7.26801 22.732 1 15 1C7.26801 1 1 7.26801 1 15C1 22.732 7.26801 29 15 29Z" stroke={color} />
+        <path d="M18.8889 18.65C18.8889 18.8433 18.7322 19 18.5389 19H11.4611C11.2678 19 11.1111 18.8433 11.1111 18.65V17.4621C11.1111 17.1479 10.7292 16.9928 10.5102 17.2181L8.2372 19.556C8.10513 19.6919 8.10513 19.9081 8.2372 20.044L10.5102 22.3819C10.7292 22.6072 11.1111 22.4521 11.1111 22.1379V20.95C11.1111 20.7567 11.2678 20.6 11.4611 20.6H20.0944C20.2877 20.6 20.4444 20.4433 20.4444 20.25V16.15C20.4444 15.9567 20.2877 15.8 20.0944 15.8H19.2389C19.0456 15.8 18.8889 15.9567 18.8889 16.15V18.65ZM11.1111 12.35C11.1111 12.1567 11.2678 12 11.4611 12H18.5389C18.7322 12 18.8889 12.1567 18.8889 12.35V13.5379C18.8889 13.8521 19.2708 14.0072 19.4898 13.7819L21.7628 11.444C21.8949 11.3081 21.8949 11.0919 21.7628 10.956L19.4898 8.61812C19.2708 8.39284 18.8889 8.5479 18.8889 8.8621V10.05C18.8889 10.2433 18.7322 10.4 18.5389 10.4H9.90556C9.71226 10.4 9.55556 10.5567 9.55556 10.75V14.85C9.55556 15.0433 9.71226 15.2 9.90556 15.2H10.7611C10.9544 15.2 11.1111 15.0433 11.1111 14.85V12.35Z" fill={color} />
```
[the remark this answers is not in the document]
Comment (whymarrh) [reply ui/app/components/ui/icon/icon.component.js]: I wouldn't suggest breaking up the `d` prop, if that's what you're suggesting. Putting each prop on its own line, sure, but I don't we should split the props.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/components/ui/icon/icon.component.js]: Yep, that is what I was suggesting. Why not split it up? We've done it elsewhere already with inlined SVGs. It's strictly more readable - the horizontal scroll here is _insane_

**Review by rachelcope — DISMISSED** [review]
Looks great! 🚀

File: ui/app/components/ui/icon/approve-icon.component.js [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    />
+    <path
+      d="M5.34426 16.0923C5.15708 16.2694 5.15656 16.5672 5.34311 16.745L9.49708 20.7032C9.67134 20.8692 9.94541 20.8687 10.1191 20.7021L10.682 20.1619C10.867 19.9844 10.8665 19.6883 10.6808 19.5114L6.53084 15.557C6.35747 15.3918 6.08509 15.3914 5.91113 15.556L5.34426 16.0923ZM24.0891 10.2959C23.9152 10.1303 23.6419 10.1303 23.4681 10.2961L14.9882 18.3839C14.8143 18.5498 14.5407 18.5497 14.3668 18.3837L11.4072 15.5567C11.2343 15.3916 10.9625 15.3905 10.7882 15.5542L10.2154 16.0924C10.0272 16.2692 10.0261 16.5679 10.2131 16.7461L14.367 20.7042C14.5408 20.8698 14.814 20.8698 14.9878 20.7042L24.6581 11.4897C24.8442 11.3124 24.8442 11.0155 24.6581 10.8382L24.0891 10.2959ZM19.7905 11.4886C19.9761 11.3117 19.9767 11.0156 19.7916 10.8381L19.2288 10.2979C19.0551 10.1313 18.781 10.1308 18.6068 10.2968L13.799 14.878C13.6125 15.0557 13.613 15.3535 13.8002 15.5306L14.367 16.067C14.541 16.2316 14.8134 16.2311 14.9868 16.0659L19.7905 11.4886Z"
+      fill={color}
+    />
+  </svg>
+)
+
+Approve.defaultProps = {
+  className: 'approve__icon',
```
Comment (Gudahtt) [inline ui/app/components/ui/icon/approve-icon.component.js]: I don't think we should be assigning classnames that aren't in-use. A better default might be to leave it without any classname
```suggestion
  className: undefined,
```

Comment (Gudahtt) [reply ui/app/components/ui/icon/approve-icon.component.js]: The same goes for the other icons as well

File: ui/app/components/ui/icon/approve-icon.component.js [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+Approve.propTypes = {
+  className: PropTypes.string,
+  width: PropTypes.number,
```
Comment (Gudahtt) [inline ui/app/components/ui/icon/approve-icon.component.js]: Does it make sense for there to be independent `width` and `height` props? Setting one larger than the other basically has no effect. A single "size" prop might make more sense.

File: ui/app/components/ui/icon/icon.stories.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+export const approveSpendLimit = () => (
+  <Approve
+    width={number('width', 28)}
+    height={number('height', 28)}
+    color={color('color', '#2F80ED')}
+  />
+)
+
+export const preloader = () => (
+  <Preloader
```
Comment (Gudahtt) [inline ui/app/components/ui/icon/icon.stories.js]: Huh - no spinning 😕 
I think the problem is that the spinning is in the `list-item` styles. Could you move into into a separate `.scss` file specifically for this icon? I would think that it should spin the same way everywhere it's used.

File: ui/app/components/ui/popover/index.scss (hunk 19 above)
Comment (Gudahtt) [reply ui/app/components/ui/popover/index.scss]: These styles still appear to be present (the popup doesn't have any `advanced-input-controls`)

**Comment by rickycodes** [comment]
Closing this in favour of several smaller bite-sized PRs like: https://github.com/MetaMask/metamask-extension/pull/8180
